### PR TITLE
Warn when truncated sampling biases the vLLM importance-sampling ratio

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
GRPO can compare vLLM sampling log probabilities with a different training distribution and silently bias its correction or off-policy mask. I added configuration warnings for sampling settings and overrides that cause this mismatch, including very small positive temperatures. The warning tests pass, and a GPU run confirmed the predicted bias from truncated sampling.
Refs #6789.

---

> [!NOTE]
> **Low Risk**
> Behavior change is advisory warnings at config init only; training math is unchanged, though some valid configs may see new UserWarnings.
> 
> **Overview**
> Adds a **`GRPOConfig` initialization warning** when vLLM is on and either **importance-sampling correction** or an **off-policy mask** will use vLLM’s returned logprobs while sampling settings reshape the distribution relative to the trainer’s full-vocab log-softmax.
> 
> The check uses **effective** `top_p`, `top_k`, `min_p`, `repetition_penalty`, and `temperature` (including **`generation_kwargs` overrides** and vLLM’s **0.01 temperature floor**). It stays **silent** without vLLM, when nothing consumes those logprobs, or when sampling stays at the documented defaults.
> 
> **Docs** for `temperature` and `generation_kwargs` now explain this bias and what is (and is not) validated. **Tests** cover the parametrized sampling cases, override behavior, off-policy mask with correction off, and no-warning paths.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 383f2160e538583bbb63977e0d942869f9509e92. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by verma8076** [comment]
Nice work — the measured-vs-predicted `|log S|` table is exactly the right way to pin this down, and scoping to a warning (rather than guessing at the truncation mask) is the right call given vLLM's kept-token set isn't exposed until #49577, which is past the current pin.

One gap: this only guards `GRPOConfig`, so `AsyncGRPOTrainer` gets no warning at all. `AsyncGRPOConfig` doesn't subclass `GRPOConfig` — it extends `_BaseConfig` directly — but it has the same `top_p`/`top_k`/`min_p` fields, and `async_grpo_trainer.py`'s `log_ratio = log_probs - old_log_probs` (line 1113) is the same mechanism: `old_log_probs` is collated straight from the rollout's vLLM logprobs (collator, line 572) with no `vllm_importance_sampling_correction`-style toggle to opt out of. So the bias isn't just uncovered by this warning there, it's unconditional — every AsyncGRPO run with truncated sampling has a biased `coef_1 = torch.exp(log_ratio)` feeding the clipped loss, with nothing to disable and no message pointing at why.

Given AsyncGRPO is arguably the more exposed case (no correction flag to turn off), might be worth either duplicating this guard into `AsyncGRPOConfig.__post_init__` (if it has one) or at least noting the scope gap explicitly in the PR description so it doesn't read as "all vLLM-backed RL trainers covered." Happy to help wire that up if useful.

**Comment by behroozazarkhalili** [comment]
Confirmed, all three, and the gap turns out to be worse than a missing warning.

Checked against this PR's head and vLLM 0.22.0:

- `AsyncGRPOConfig` is at `async_grpo_config.py:22` and `GRPOConfig` at `grpo_config.py:23`, both extending `_BaseConfig` directly. They are siblings, so nothing is inherited.
- `top_p:241`, `top_k:248`, `min_p:255`.
- `log_ratio` at `async_grpo_trainer.py:1113`, `old_log_probs` collated at `:559-572`, and `grep importance_sampling` returns nothing in either async file.

You hedged on whether it has a `__post_init__`. It does, at `:384`, so there was somewhere to hang the guard.

What I had not appreciated until working through it is that the async bias lands in a different place than GRPO's. GRPO computes its own `old_per_token_logps` (`grpo_trainer.py:2664`) and differences it against vLLM's, so the bias sits on a correction term that `vllm_importance_sampling_correction=False` switches off. AsyncGRPO uses the vLLM logprob as the PPO denominator itself, so for an unchanged policy `coef_1` reads the surviving mass instead of 1.0:

| `top_p` | `coef_1` | error vs 1.0 | outside `(0.8, 1.2)`? |
| --- | --- | --- | --- |
| 1.00 | 1.0000 | 0.0000 | no |
| 0.90 | 0.9000 | 0.1000 | no |
| 0.80 | 0.8000 | 0.2000 | no |
| 0.70 | 0.7000 | 0.3000 | yes |

Below `1 - epsilon_low` the clip engages on every negative-advantage token whether or not the policy moved.

Two things came out of it. #6944 adds the warning on the AsyncGRPO side, and its CI is green. #6945 raises the design question underneath, which is that TRL imposes `--logprobs-mode processed_logprobs` on every server it prescribes and justifies it by the importance-sampling correction, a correction AsyncGRPO does not have.

Your comment also caught something in this PR itself. `repetition_penalty` reaches the same logprobs, because penalties are applied at `sampler.py:403` before `sample()` computes them, and the guard here listed only the three truncation knobs. Fixed in 5107c7df with a test.

On the offer to help, the useful piece is measurement. That table is derived from the two normalizations and computed exactly, not observed on hardware, and confirming it end to end needs a two-GPU async run against a real vLLM server that I have not managed to schedule. If you have a lane for that, #6945 is the place to put it.

File: trl/trainer/grpo_config.py [hunk 1]
```diff
@@ -1134,6 +1137,68 @@ def __post_init__(self):
...
+                    "and the off-policy mask: vLLM computes the logprobs it returns after the sampler has applied "
+                    "these settings, while the trainer normalizes over the full vocabulary and divides by the "
+                    "configured `temperature` only, so their difference includes that reshaping rather than the "
+                    "train/inference mismatch alone. Keep the sampling defaults (`top_p=1.0`, `top_k=0`, "
+                    "`min_p=None`, `repetition_penalty=1.0`, `temperature` not overridden) until the correction "
+                    "accounts for it, or disable both consumers with `vllm_importance_sampling_correction=False` and "
+                    "`off_policy_mask_threshold=None`. See https://github.com/huggingface/trl/issues/6789.",
+                    UserWarning,
+                    stacklevel=3,
+                )
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/trainer/grpo_config.py:1220]: Correct that `AsyncGRPOConfig` does not inherit this block. That is deliberate: the async trainer has no `vllm_importance_sampling_correction`, so the message here ("biases `sampling/sampling_logp_difference` and the correction built on it") would be false for it, and its own guard needs different wording. That half is #6944, which adds the warning to `AsyncGRPOConfig.__post_init__` with the async-specific message and covers `top_p`, `top_k`, `min_p` and `repetition_penalty`; #6945 tracks the design question of whether AsyncGRPO should get the correction itself. Keeping the two PRs separate lets each be reviewed against its own trainer's semantics.
